python/ray/_private/ray_logging.py

```diff
@@ -32,14 +30,6 @@ def setup_logger(
     if type(logging_level) is str:
         logging_level = logging.getLevelName(logging_level.upper())
     logger.setLevel(logging_level)
-    global _default_handler
```

Comment:
The issue may happen here. 

```python
# Comments in the function `_setup_logging()`
    # Logs will also be written to the container's stdout.
    # The stdout handler was set up in the Ray CLI entry point.
    # See ray.scripts.scripts::cli().
```

[run_kuberay_autoscaler()](https://sourcegraph.com/github.com/ray-project/ray@ac2230fd28f642e4523422a1e1eb15ab92f63943/-/blob/python/ray/autoscaler/_private/kuberay/run_autoscaler.py?L19:5-19:27) -> [_setup_logging()](https://sourcegraph.com/github.com/ray-project/ray@ac2230fd28f642e4523422a1e1eb15ab92f63943/-/blob/python/ray/autoscaler/_private/kuberay/run_autoscaler.py?L67:5-67:19) -> [cli()](https://sourcegraph.com/github.com/ray-project/ray@ac2230fd28f642e4523422a1e1eb15ab92f63943/-/blob/python/ray/scripts/scripts.py?L74) -> [setup_logger()](https://sourcegraph.com/github.com/ray-project/ray@ac2230fd28f642e4523422a1e1eb15ab92f63943/-/blob/python/ray/_private/ray_logging.py?L14:5-14:17)

```python
# setup_logger
    if _default_handler is None:
        _default_handler = logging.StreamHandler()
        logger.addHandler(_default_handler)
```

> The [StreamHandler](https://docs.python.org/3/library/logging.handlers.html#logging.StreamHandler) class, located in the core [logging](https://docs.python.org/3/library/logging.html#module-logging) package, sends logging output to streams such as sys.stdout, sys.stderr or any file-like object (or, more precisely, any object which supports write() and flush() methods). (ref: https://docs.python.org/3/library/logging.handlers.html)

Reply:
Hmm the bit about `cli()` does seem suspicious. But in this branch, a `_StderrHandler` is attached to the `ray` logger in `ray/python/ray/_private/log.py` whenever `ray` is imported now, and `setup_logger` doesn't set up any handlers anymore.

Reply:
Maybe the comment from the function `_setup_logging()` is a bit wrong. The default stream of StreamHandler is `stderr` rather than `stdout`. In addition, `kubectl logs` will collect both stdout and stderr ([ref](https://kubernetes.io/docs/concepts/cluster-administration/logging/)).

```python
# test.py
import logging

_default_handler = None

def get_logger(name):
    logger = logging.getLogger(name)
    global _default_handler
    if _default_handler is None:
        _default_handler = logging.StreamHandler()
        logger.addHandler(_default_handler)
    return logger

logger = get_logger("my_logger")
logger.setLevel(logging.INFO)
logger.info("Hello, World!")
```

* `python3 test.py | tee log` => `cat log` => Nothing (nothing in STDOUT)
* `python3 test.py 2>&1 | tee log` => `cat log` => "Hello, World!"

Reply:
Could we use `print(logger.handlers)` in `setup_logging()` to check the attached handlers?

Reply:
You're right, the `StreamHandler` defaults to `sys.stderr`. Interestingly, the operator test [fails here](https://github.com/peytondmurray/ray/blob/b5db67b1248c4f67c0e76c887adcfd5039e740f9/python/ray/tests/kuberay/test_autoscaling_e2e.py#L253), when calling `scale_up.py`. And if I print the loggers and their respective handlers inside `scale_up.py` I get the following:

```
~~~~~~~~~~~~~~~scale_up.py~~~~~~~~~~~~~~
<RootLogger root (WARNING)>
asyncio
charset_normalizer
  <StreamHandler <stderr> (NOTSET)>
concurrent
concurrent.futures
filelock
grpc
  <NullHandler (NOTSET)>
grpc._channel
grpc._common
grpc._cython
grpc._cython.cygrpc
grpc._simple_stubs
grpc.aio
grpc.aio._call
pkg_resources
pkg_resources.extern
pkg_resources.extern.packaging
pkg_resources.extern.packaging.tags
ray
ray._private
ray._private.external_storage
ray._private.function_manager
ray._private.gcs_pubsub
ray._private.gcs_utils
ray._private.import_thread
ray._private.memory_monitor
ray._private.node
ray._private.parameter
ray._private.ray_constants
ray._private.resource_spec
ray._private.runtime_env
ray._private.runtime_env.conda
ray._private.runtime_env.context
ray._private.runtime_env.packaging
ray._private.runtime_env.pip
ray._private.runtime_env.plugin
ray._private.runtime_env.plugin_schema_manager
ray._private.runtime_env.py_modules
ray._private.runtime_env.uri_cache
ray._private.runtime_env.validation
ray._private.runtime_env.working_dir
ray._private.serialization
ray._private.services
ray._private.signature
ray._private.state
ray._private.test_utils
ray._private.usage
ray._private.usage.usage_lib
ray._private.utils
ray._private.worker
ray._raylet
ray.actor
ray.client_builder
ray.cloudpickle
ray.cloudpickle.compat
ray.experimental
ray.experimental.tqdm_ray
ray.remote_function
ray.runtime_context
ray.runtime_env
ray.runtime_env.runtime_env
ray.util
ray.util.client
ray.util.client.api
ray.util.client_connect
ray.util.rpdb
ray.util.tracing
ray.util.tracing.tracing_helper
ray.widgets
ray.widgets.util
requests
  <NullHandler (NOTSET)>
socks
tqdm
tqdm.cli
urllib3
  <NullHandler (NOTSET)>
urllib3.connection
urllib3.connectionpool
urllib3.poolmanager
urllib3.response
urllib3.util
urllib3.util.retry
```

Importantly, there seems to be no handlers set for `ray` (because `setup_component_logger` was called?), which explains why no error messages appear.

Also, it looks like `scale_up.py` really is searching for log messages on `stdout`, which may or may not be related to the messages not showing up at all.